src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php

```diff
@@ -0,0 +1,156 @@
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\PropertyInfo;
+
+use Doctrine\Common\Persistence\Mapping\ClassMetadataFactory;
+use Doctrine\Common\Persistence\Mapping\MappingException;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use Symfony\Component\PropertyInfo\PropertyListRetrieverInterface;
+use Symfony\Component\PropertyInfo\PropertyTypeInfoInterface;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Extracts data using Doctrine ORM and ODM metadata.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class DoctrineExtractor implements PropertyListRetrieverInterface, PropertyTypeInfoInterface
+{
+    /**
+     * @var ClassMetadataFactory
+     */
+    private $classMetadataFactory;
+
+    public function __construct(ClassMetadataFactory $classMetadataFactory)
+    {
+        $this->classMetadataFactory = $classMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getProperties($class, array $context = array())
+    {
+        try {
+            $metadata = $this->classMetadataFactory->getMetadataFor($class);
+        } catch (MappingException $exception) {
+            return;
+        }
+
+        return array_merge($metadata->getFieldNames(), $metadata->getAssociationNames());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTypes($class, $property, array $context = array())
+    {
+        try {
+            $metadata = $this->classMetadataFactory->getMetadataFor($class);
+        } catch (MappingException $exception) {
+            return;
+        }
+
+        if ($metadata->hasAssociation($property)) {
+            $class = $metadata->getAssociationTargetClass($property);
+
+            if ($metadata->isSingleValuedAssociation($property)) {
+                if ($metadata instanceof ClassMetadataInfo) {
+                    $nullable = isset($metadata->discriminatorColumn['nullable']) ? $metadata->discriminatorColumn['nullable'] : false;
+                } else {
+                    $nullable = false;
+                }
+
+                return array(new Type(Type::BUILTIN_TYPE_OBJECT, $nullable, $class));
+            }
+
+            return array(new Type(
+                Type::BUILTIN_TYPE_OBJECT,
+                false,
+                'Doctrine\Common\Collections\Collection',
+                true,
+                new Type(Type::BUILTIN_TYPE_INT),
+                new Type(Type::BUILTIN_TYPE_OBJECT, false, $class)
+            ));
+        }
+
+        if ($metadata->hasField($property)) {
+            $typeOfField = $metadata->getTypeOfField($property);
+            if ($metadata instanceof ClassMetadataInfo) {
+                $nullable = $metadata->isNullable($property);
+            } else {
+                $nullable = false;
```

Comment:
Unnecessary else. You can either move `$nullable = false;` above the `if` block or simplify the whole block to:
```php
$nullable = $metadata instanceof ClassMetadataInfo && $metadata->isNullable($property);
```